[clang] [libc++] fix _Atomic c11 compare exchange does not update expected results

huixie90 opened the pull request:
fixes #30023

The issue is that for compare exchange builtin, if the type's size is not power of 2, it creates a temporary of size power of 2, then emit the compare exchange operation. And later, the results of the compare exchange operation has two components: 1. a boolean whether or not the exchange happens. 2. the old value
we are supposed to write the old value into user's "expected" value. However, in case the type is not power of 2, what we actually wrote to is the temporary that was created.

The fix is to pass the "expected" address all the way down so it can wrote to the correct address

huixie90 committed:
[libc++] fix _Atomic c11 compare exchange does not update expected results

ldionne commented on clang/lib/CodeGen/CGAtomic.cpp:
```diff
@@ -408,7 +409,15 @@ static void emitAtomicCmpXchg(CodeGenFunction &CGF, AtomicExpr *E, bool IsWeak,
+  if (OriginalSizeInBits / 8 == Size) {
+    CGF.Builder.CreateStore(Old, ExpectedResult);
+  } else {
+    // How to just store N bytes to ExpectedResult ?
+    CGF.Builder.CreateStore(Old, ExpectedResult);
```
@bcardosolopes Hi Bruno! I'm chatting with @huixie90 right now and it seems like he needs to store an N bytes temporary object (that was used to perform the compare exchange above, but where only the first M bytes are used) into the actual result, which has exactly M bytes. We're novices in LLVM IR here -- can you suggest someone that would be able to answer a few questions like this one? Starting with you since Github suggested you as a reviewer for having modified these files recently :)

huixie90 committed:
update

efriedma-quic commented on clang/lib/CodeGen/CGAtomic.cpp:
```diff
@@ -411,7 +412,48 @@ static void emitAtomicCmpXchg(CodeGenFunction &CGF, AtomicExpr *E, bool IsWeak,
+  llvm::Type *I8PtrTy = llvm::PointerType::getUnqual(CGF.getLLVMContext());
+
+  llvm::Value *SrcPtr = CGF.Builder.CreateBitCast(OldStorage.getBasePointer(), I8PtrTy);
+  llvm::Value *DstPtr = CGF.Builder.CreateBitCast(ExpectedResult.getBasePointer(), I8PtrTy);
```
Bitcasts don't do anything with opaque pointers.

efriedma-quic commented on clang/lib/CodeGen/CGAtomic.cpp:
```diff
@@ -411,7 +412,48 @@ static void emitAtomicCmpXchg(CodeGenFunction &CGF, AtomicExpr *E, bool IsWeak,
+  llvm::Align DstAlignLLVM = DL.getABITypeAlign(ExpectedType);
+
+  clang::CharUnits SrcAlign = clang::CharUnits::fromQuantity(SrcAlignLLVM.value());
+  clang::CharUnits DstAlign = clang::CharUnits::fromQuantity(DstAlignLLVM.value());
```
This is not the correct way to query alignment in general.

An Address should already have alignment encoded in it.

huixie90 replied:
Since `Old` is the result of the compare exchange. I assume it has the same alignment of the atomic address itself. so I used the `Ptr.getAlignment()` here. would that be fine?

efriedma-quic commented on clang/lib/CodeGen/CGAtomic.cpp:
```diff
@@ -411,7 +412,48 @@ static void emitAtomicCmpXchg(CodeGenFunction &CGF, AtomicExpr *E, bool IsWeak,
+  clang::CharUnits DstAlign = clang::CharUnits::fromQuantity(DstAlignLLVM.value());
+
+  // Allocate temporary storage for Old value
+  llvm::AllocaInst *Alloca = CGF.CreateTempAlloca(OldType, "old.tmp");
```
I think there's a variant of CreateTempAlloca that returns an Address?

huixie90 committed:
address feedback

ldionne reviewed (commented):
This looks quite reasonable to me, and it seems we're fixing a pretty bad bug. However, someone else with more Clang/Codegen experience needs to look at this. Thanks for working on this!

CC @efriedma-quic

efriedma-quic commented:
I think the current state addresses my concerns, yes.  I'd like to make one final pass over it, though; please git-clang-format and rebase the patch.

huixie90 committed:
rebase

huixie90 committed:
clang format

huixie90 committed:
fix c atomic test

huixie90 commented:
> I think the current state addresses my concerns, yes. I'd like to make one final pass over it, though; please git-clang-format and rebase the patch.

Thanks I rebased and clang-formatted the changed lines.  The CI fails on the new test I added in libc++, because the libc++ is not using the just-built clang for testing for majority of the targets and they fail as expected. i will check with @ldionne  how to skip the previous versions of clangs for that test

huixie90 committed:
fix temporary not zeroed

efriedma-quic reviewed (approved):
LGTM

huixie90 committed:
disable the test for clang-22 until the fix is included in libc++ CI image

huixie90 committed:
c++11 ci fix

huixie90 committed:
clang-19 for free bsd
